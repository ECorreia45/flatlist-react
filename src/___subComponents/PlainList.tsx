import {array, func, node, object, oneOfType, string} from 'prop-types';
import React, {forwardRef, memo, Ref} from 'react';
import arePropsEqual from '../___utils/arePropsEqual';
import convertListToArray from '../___utils/convertListToArray';
import {isString} from '../___utils/isType';
import DefaultBlank from './DefaultBlank';
import {DisplayHandlerProps} from './DisplayHandler';
import {handleRenderItem, renderBlank, renderFunc} from './uiFunctions';

interface Props<T> extends DisplayHandlerProps {
    list: T[];
    renderItem: JSX.Element | renderFunc;
    renderWhenEmpty: null | (() => JSX.Element);
    wrapperHtmlTag: string;
    __forwarededRef: object;
}

// const List = PlainList() as ForwardRefExoticComponentExtended;

const PlainList = (props: Props<DisplayHandlerProps | []>) => {
    const {
        list, renderItem, renderWhenEmpty, wrapperHtmlTag, __forwarededRef,
        ...tagProps
    } = props;
    const renderList = convertListToArray(list);

    if (renderList.length === 0) {
        return renderBlank(renderWhenEmpty);
    }

    const renderThisItem = handleRenderItem(renderItem);

    const WrapperElement = `${isString(wrapperHtmlTag) && wrapperHtmlTag ? wrapperHtmlTag : ''}`;
    const content = (
        <>
            {
                renderList.length > 0
                    ? renderList.map(renderThisItem)
                    : renderBlank(renderWhenEmpty)
            }
        </>
    );

    return (
        <>
            {
                WrapperElement
                    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
                    // @ts-ignore
                    ? <WrapperElement {...tagProps} ref={__forwarededRef}>{content}</WrapperElement>
                    : content
            }
        </>
    );
};

PlainList.propTypes = {
    /**
     * a list of anything to be displayed
     */
    list: oneOfType([array, object]).isRequired,
    /**
     * a jsx element or a function that it is called for every item on the list and returns a jsx element
     */
    renderItem: oneOfType([func, node]).isRequired,
    /**
     * the function that gets called when the list is empty or was filtered to the point it became empty
     */
    renderWhenEmpty: func,
    /**
     * a optional html tag to use to wrap the list items
     */
    wrapperHtmlTag: string,
    // eslint-disable-next-line react/forbid-prop-types
    __forwarededRef: object
};

PlainList.defaultProps = {
    wrapperHtmlTag: '',
    renderWhenEmpty: DefaultBlank,
    __forwarededRef: {}
};

export default memo(forwardRef((props: Props<DisplayHandlerProps | []>, ref: Ref<HTMLElement>) => (
    <PlainList __forwarededRef={ref} {...props} />
)), arePropsEqual);
