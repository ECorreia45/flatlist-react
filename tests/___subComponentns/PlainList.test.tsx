import React from 'react';
import PlainList from '../../src/___subComponents/PlainList';
import {render} from '@testing-library/react';

describe('PlainList', () => {
    const list = [
        {name: 'item-1'},
        {name: 'item-2'},
        {name: 'item-3'}
    ];

    it('Should render items', () => {
        const {asFragment, getAllByText} = render(
            <PlainList
                list={list}
                renderItem={(item: any, k: any) => <li key={k}>{item.name}</li>}
            />
        );

        const items = getAllByText(/item-*/);

        expect(asFragment()).toMatchSnapshot();
        expect(items.length).toBe(3);
        expect(items.map(item => item.textContent)).toEqual(['item-1', 'item-2', 'item-3']);
    });

    it('Should render items wrapped in a tag', () => {
        const {asFragment, container} = render(
            <PlainList
                list={list}
                renderItem={(item: any, k: any) => <li key={k}>{item.name}</li>}
                wrapperHtmlTag="div"
                id="container"
            />
        );

        expect(asFragment()).toMatchSnapshot();
        expect(container.children[0].id).toBe('container');
    });

    it('Should render blank with empty list', () => {
        const {asFragment, getByText} = render(<PlainList list={[]} renderItem={() => null}/>);

        expect(asFragment()).toMatchSnapshot();
        expect(getByText('List is empty...').outerHTML).toBe('<p>List is empty...</p>');
    });

    it('Should render provided blank with empty list', () => {
        const {asFragment, getByText} = render(<PlainList list={[]} renderItem={() => null}
                                                          renderWhenEmpty={() => <div>Empty</div>}/>);

        expect(asFragment()).toMatchSnapshot();
        expect(getByText('Empty').outerHTML).toBe('<div>Empty</div>');
    });
});
