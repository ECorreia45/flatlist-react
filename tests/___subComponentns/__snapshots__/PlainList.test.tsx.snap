// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlainList Should render blank with empty list 1`] = `
<DocumentFragment>
  <p>
    List is empty...
  </p>
</DocumentFragment>
`;

exports[`PlainList Should render items 1`] = `
<DocumentFragment>
  <li>
    item-1
  </li>
  <li>
    item-2
  </li>
  <li>
    item-3
  </li>
</DocumentFragment>
`;

exports[`PlainList Should render items wrapped in a tag 1`] = `
<DocumentFragment>
  <div
    id="container"
  >
    <li>
      item-1
    </li>
    <li>
      item-2
    </li>
    <li>
      item-3
    </li>
  </div>
</DocumentFragment>
`;

exports[`PlainList Should render provided blank with empty list 1`] = `
<DocumentFragment>
  <div>
    Empty
  </div>
</DocumentFragment>
`;
